contributing/architecture-eval-lifecycle.md

```diff
@@ -0,0 +1,326 @@
+scheduler executes 3 main steps:
+
+* Reconcile: compare the cluster state and job specification to determine what
+  changes need to be made -- starting or stopping Allocations. The scheduler
+  creates new Allocations in this step. But note that _these Allocations are not
+  yet persisted to the state store_.
+* Feasibility Check: For each Allocation it needs, the scheduler iterates over
+  Nodes until it finds 2 Nodes that match the Allocation's resource requirements
+  and constraints.
+  * Note: for system jobs or jobs with with `spread` blocks, the scheduler has
+    to check all Nodes.
+* Scoring: For each feasible node, the scheduler ranks them and picks the one
+  with the highest score.
+
+If the job is a `service` job, the scheduler will also create (or update) a
+Deployment. When the scheduler determines the number of Allocations to create,
+it examine the job's [`update`][] block. Only the number of Allocations needed
+to complete the next phase of the update will be created in a given
+allocation. The Deployment is used by the deployment watcher (see below) to
+monitor the health of allocations and create new Evaluations to continue the
+update.
+
+If the scheduler cannot place all Allocations, it will create a new Evaluation
+in the "blocked" state and submit it to the leader. The Eval Broker will
+re-enque that Evaluation once cluster state has changed. (This process is the
+first green box in the sequence diagram below.)
+
+Once the scheduler has completed processing the Evaluation, if there are
+Allocations (and possibly a Deployment) to update, it will submit this work as a
+Plan to the leader. The leader needs to validate this plan and serialize it:
+
+* The scheduler took a snapshot of cluster state at the start of its work, so
+  that state may have changed in the meantime.
+* Schedulers run concurrently across the cluster, so they may generate
+  conflicting plans (particularly on heavily-packed clusters).
+
+The leader processes the plan in the plan applier. If the plan is valid, the
+plan applier will write the Allocations (and Deployment) update to the state
+store. If not, it will reject the plan. (The plan submit process is the second
+green box in the sequence diagram below.)
+
+Once the scheduler has a response from the leader, it will tell the Eval Broker
+to Ack the Evaluation (if it successfully submitted the plan) or Nack the
+Evaluation (if it failed to do so).
+
+The diagram below shows the scheduling phase, including submitting plans to the
+planner. Note that at the end of this phase, Allocations (and Deployment) have
+been persisted to the state store.
+
+```mermaid
+sequenceDiagram
+
+    participant leaderRpc as Leader RPC
+    participant leaderFSM as Leader State Store
+    participant planner as Planner
+    participant broker as Eval Broker
+    participant workerFSM as Worker State Store
+    participant sched as Scheduler
+
+    broker ->> leaderFSM: blocking query for new evals
```

Comment:
Is it a blocking query or do evals get enqueued directly using [`EvalBroker.Enqueue`](https://github.com/hashicorp/nomad/blob/main/nomad/eval_broker.go#L189-L194)?

Reply:
Good catch, I was confusing the [restoreEvals](https://github.com/hashicorp/nomad/blob/v1.4.0-beta.1/nomad/leader.go#L571-L597) that happens on leader transition with [upserting evals](https://github.com/hashicorp/nomad/blob/v1.4.0-beta.1/nomad/leader.go#L571-L597). Will fix!

Reply:
Done